🌟 Implement a SearchEngine for quick in-memory searching of Time Entries

kornelijepetak opened the pull request:
## What's this?
Implements caching and in-memory search of the time entries to speed up the typing in the Start TE view.

⚠ Please read the whole PR description before looking at the code.

### Relationships
Closes #5473

## Why do we want this?
For speed.

## How is it done?
By 
* fetching data in memory
* caching filtered time entries for various used query texts
* optimizing changes (appending and prepending)

I've used the `Task<IEnumerable<T>>` as a return type, but I'm thinking whether it maybe makes sense to use `IObservable<IEnumerable<T>>` instead. What do you think?

### Why not another way?
That's where you come in.

### Side effects
There's a thing. I have a couple thousand time entries on the phone. Now, it seems that there's still a small delay and it seems to be related to the refresh/redraw of the UI after the change.

However, it feels like the opening of the Start TE now happens with a small delay. I'm not sure if I'm maybe initializing the inital data on the UI thread?

## :squid: Permissions
Me, @heytherewill or @amulware

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+        public async Task<IEnumerable<T>> Get(string query)
+        {
+            await initialDataAvailable.Task;
+            return await Task.Run(() => get(query).AsEnumerable());
```
There's a lot of accesses to the cache in get(), but nothing to ensure thread safety. What will happen if a user types faster than we can search, or other threading stuff happens?

kornelijepetak replied:
Thread safety concerns are a valid point. I will take it into account.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+            var (dataSource, changeType) = determineBaseDataSource(query);
+
+            if (changeType == Append || changeType == Prepend)
+                cache[query] = dataSource;
```
Why do we store the base data source for the new query? That doesn't seem right.

kornelijepetak replied:
Yes, in addition to your next comment, this is a bug. I've been reworking this method a lot and it ended up wrong. Fixing it.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+            if (changeType == Append || changeType == Prepend)
+                cache[query] = dataSource;
+
+            return dataSource.Filter(query).Items;
```
We should be storing the filtered one here instead, right? Instead of forgetting it?

kornelijepetak replied:
See previous comment.

amulware commented on Toggl.Core/SearchBox/SearchEngineCacheItem.cs:
```diff
@@ -0,0 +1,39 @@
+{
+    internal class SearchEngineCacheItem<T>
+    {
+        private readonly Func<string, T, bool> filter;
```
It seems very strange to me that we store this filter here, it's not really needed for anything, is it? We could just inject it in Filter()?

kornelijepetak replied:
Yes, that could work. Less memory.

amulware replied:
Also semantically better I think.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+            return await Task.Run(() => get(query).AsEnumerable());
+        }
+
+        private ImmutableList<T> get(string query)
```
Why is there no shortcut in here to return the cached list if it's a known query?

kornelijepetak replied:
How is there not?
The first thing that `determineBaseDataSource()` does is check for existance of the cached entry.

amulware replied:
But there is no early return in get() it will still do a Filter() call, which is the most expensive thing in the whole algorithm.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+                return (dataSource, Prepend);
+
+            // No optimization found, use all 
+            return (completeData, Other);
```
We can remove this case and the `completeData` field if we add the `completeData` cache item to the dictionary. Maybe worth simplifying? Or is there a benefit in telling apart the 'other' case for something?

kornelijepetak replied:
`completeData` should be returned for all queries that do not exist in the dictionary (and for whose there's no append/prepend combination in the dictionary as well). The number of those is infinite. I don't really understand how you mean to add such case to a dictionary. What kind of key would that be?

amulware replied:
If you add it to the dictionary for they "" key, then any other query will be an extension of it in the Append/Prepend case.

kornelijepetak replied:
Oh, yes, you are right. 
But it does introduce a layer of questioning when reading the code, because these are implicit, but I've tried to signify that in the exception message. Not sure if it's any clearer though.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,77 @@
+
+            // Character preppended to a previously known query
+            dataSource = cache
+                .OrderByDescending(cachedItem => cachedItem.Key.Length)
```
This is the same enumeration as just above, could we maybe do them both in the same loop instead?

amulware replied:
Also, since we're always enumerating the cache sorted by size, should we maybe just keep it ordered and use a different collection for that?

kornelijepetak replied:
I changed this to a single loop.

However, using any other collection seems like an overkill (if there is such a collection at all). AFAIK, there's no collection that can handle key-value pairs and be sorted by something else than the same key. Because that's what we would need. Keys need to be queries, but the ordering is done on the query lengths. 
The only 2 ways I see it working is `Dictionary<int, Dictionary<string, DataSource>>` (which then has a problem of unwrapping for direct access) or a combination of two different collection (a dictionary to enable O(n) access, and a `SortedList` to work with fast enumeration). Both seem like even a more complex idea.

Especially since this part of the code is not even performance critical. The dictionary search (even with two loops as before) lasts less than a few `ms`. I don't think it's worthwile introducing more complex structures, let alone writing our own.

amulware replied:
There is SortedSet<T>, but I don't know right now how you'd best implement the lookup for the exact match (binary + linear search I guess, not great..), though you could use both a sorted set and a dictionary. But I agree that it might not be worth the complication (though I generally don't like OrderBy calls when we're only retrieving a single item in the end).

kornelijepetak replied:
For now I won't introduce any more complexity.
Also, the performance of this should always be fine, since we should probably limit the number of cached items. Otherwise, it could lead to some memory problems. Not that fast, but still.
How do you suggest we approach the limit?

amulware replied:
This will be recreated everytime we open the startview, right? I.e. the cache is pretty shortlived.. I don't know if we really care about limiting it?

kornelijepetak replied:
Yes. It is short lived, unless somebody really likes our Start TE view and wants to type into the filter for hours on end.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,80 @@
+
+                    if (query.EndsWith(cachedItem.Key))
+                        return (cachedItem.Value, Prepend);
+                }
```
A lot of work in this lock. Any way to simplify this? Since you're ordering anyways (which enumerates the items into a list internally, maybe enumerate them into a list manually, and then sort (maybe with a list.sort or whatever there is) and search outside the lock?

amulware replied:
Also, do we care about the difference between startswith/endswith? We could detect the key as substring instead?

kornelijepetak replied:
I replaced this with the `Contains` and ditched the enum.

amulware commented on Toggl.Core/SearchBox/SearchEngine.cs:
```diff
@@ -0,0 +1,80 @@
+
+                var sortedCacheItems = cache
+                    .OrderByDescending(cachedItem => cachedItem.Key.Length)
+                    .Where(cachedItem => cachedItem.Key.Length < query.Length);
```
Maybe switch these two lines to order fewer items, since the Where is checking all of them anyways?

kornelijepetak replied:
Switched.

kornelijepetak commented:
Tests coming soon.

kornelijepetak commented:
In several commits, I've done several changes: 

* I've had to eradicate the old interactor and its tests and fix the new tests.
* Additionally, I've moved the data initialization to the background thread because of some tests deadlocks.
* I've abstracted an implementation of the search engine for time entries. 
* I've added interface to search box for testability reasons.
* I've added tests.

@amulware @heytherewill @daividssilverio please recheck

daividssilverio commented on Toggl.Core.Tests/SearchEngine/SearchEngineTests.cs:
```diff
@@ -0,0 +1,171 @@
+            var waitPeriod = TimeSpan.FromSeconds(2);
+            var searchEngine = new SearchEngine<MockDataItem>(pass);
+
+            var hasCompleted = searchEngine.Get("").Wait(waitPeriod);
```
Hey @kornelijepetak , these `Wait(waitPeriod)` are all blocking, right? Why not make these tests async?

kornelijepetak replied:
True.

daividssilverio commented on Toggl.Core.Tests/SearchEngine/SearchEngineTests.cs:
```diff
@@ -0,0 +1,171 @@
+            var searchEngine = new SearchEngine<MockDataItem>(pass);
+            searchEngine.SetInitialData(ImmutableList<MockDataItem>.Empty);
+
+            var hasCompleted = searchEngine.Get("").Wait(waitPeriod);
```
Here as well.

kornelijepetak replied:
True.

heytherewill commented on Toggl.Core.Tests/SearchEngine/SearchEngineTests.cs:
```diff
@@ -0,0 +1,171 @@
+            var waitPeriod = TimeSpan.FromSeconds(2);
+            var searchEngine = new SearchEngine<MockDataItem>(pass);
+
+            var hasCompleted = searchEngine.Get("").Wait(waitPeriod);
```
Why not:

`var hasCompleted =await Task.WhenAny(searchEngine.Get("").ContinueWith(_ => throw new XUnitException()), Task.Delay(2)`?

heytherewill replied:
Moreover, why is this even tested? 🤔

heytherewill commented on Toggl.Core.Tests/SearchEngine/SearchEngineTests.cs:
```diff
@@ -0,0 +1,171 @@
+        [Fact]
+        public void CompletesIfInitialDataIsSet()
+        {
+            var waitPeriod = TimeSpan.FromSeconds(2);
```
Why? Also why is this test not async?

kornelijepetak replied:
🤦‍♂

kornelijepetak commented:
For logistical reasons, I'm yielding the authorship of this PR to whoever wants to complete it. I will simply not have enough time to see what's wrong with the tests.
